---
title: "GWAS Report"
output:
  html_document:
    toc: true
    theme: united
    toc_float: false
params:
  workfolder: unknown
  ident: unknown
  nr_samples: 0
  nr_markers: 0
  genoid: unknown
  lm_call: unknown
  phenoname: unknown 
  plink: unknown
  lambda: unknown 
  mac: unknown
  machr2_low: unknown 
  hwe_pval: unknown 
  excel_file: unknown
  signif_file: unknown
  cojo_done: unknown
  clump_done: unknown  
  cojo_file: unknown 
  cojo_orig_file: unknown 
  clump_file: unknown  
  cojo_neargene_file: unknown
  clump_neargene_file: unknown  
  signif_neargene_file: unknown   
  qq_plot: unknown
  man_plot: unknown
  histo_plot: unknown
  kernel_plot: unknown
---




```{r, echo = FALSE}
workfolder = params$workfolder
ident = params$ident
nr_samples = params$nr_samples
nr_markers = params$nr_markers 
genoid = params$genoid
lm_call = params$lm_call
phenoname = params$phenoname
plink = params$plink
lambda = params$lambda
mac = params$mac 
machr2_low = params$machr2_low  
hwe_pval = params$hwe_pval  
excel_file = params$excel_file
signif_file = params$signif_file
cojo_done = params$cojo_done   	
clump_done = params$clump_done 	
cojo_file = params$cojo_file 				 
cojo_orig_file = params$cojo_orig_file  		 
clump_file = params$clump_file  			 
cojo_neargene_file = params$cojo_neargene_file 		 
clump_neargene_file = params$clump_neargene_file	   
signif_neargene_file = params$signif_neargene_file 	 
qq_plot = params$qq_plot
man_plot = params$man_plot
histo_plot = params$histo_plot
kernel_plot = params$kernel_plot 
```

<br>  





## Overview

These are the GWAS results for job <span style="color:red"> `r ident` </span> 
with `r prettyNum(as.integer(nr_samples), big.mark=".", decimal.mark=",")` samples, 
run with genotype dataset **`r genoid`** for the phenotype <span style="color:red"> `r phenoname` </span>.
The genomic inflation factor was $\lambda =$ `r lambda`. 


<br>





## Methods

The association between `r phenoname` and each variant with a minor allele count $\ge$ `r mac` and 
an imputation quality metric $r^2 \ge$ `r machr2_low` was tested using the linear regression model:  

<span style="color:blue">
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `r lm_call` 
</span>

The association tests were performed using *`r plink`*, including a total of 
`r prettyNum(as.integer(nr_markers), big.mark=".", decimal.mark=",")` markers (after filtering).
Markers having a Hardy-Weinberg equilibrium exact test p-value below `r formatC(hwe_pval, format = "e", digits = 1)` were filtered out.  



<!-- prepare table content -->




```{r setup, echo = FALSE}
library(knitr)
library(DT, quietly = TRUE)
```

## Table(s)

 - Markers with $p \le 5 \cdot 10^{-8}$ are displayed, no matter what the corresponding $\beta$-value is.
 - [Spreadsheet](`r excel_file`) with significant markers (only available if copied separately)
 - ID: link to [Phenoscanner](http://www.phenoscanner.medschl.cam.ac.uk)  
 - POS: link to marker position at [UCSC Genome Brower](https://genome.ucsc.edu) 
 - genes: link to nearest gene (UCSC Genome Brower)
 - <span style="color:red"> Note </span>: *A1_FREQ*, *BETA*, *SE*, and *P* are not available if only one significant marker resides on a whole chromosome. 
 


```{r cojo_results, echo = FALSE, warning = FALSE, eval = cojo_done}
asis_output("#### GCTA-COJO results\\n")

cojo_file = paste(workfolder, cojo_file, sep ="/")  # "/castor/project/proj/GWAS_TEST/LIV_MULT4/LIV_MULT4_liv2_cojoed_markers.RData"
if(file.exists(cojo_file)) {
  results = get(load(cojo_file))
} else {
  stop(paste("\n\n  ERROR (gwas_report.Rmd): File", cojo_file, "not found.\n\n"))
}
# colnames(results)[7]  # "OBS_CT"
results = results[,-7]  # the cojoed results show effective values which are not all equal - remove anyway to stick to table format 
rownames(results) = 1:nrow(results)  

names = results[,1]                        # all marker names 
ind1 = grep("^rs", names)                  # rs names               
ind2 = grep("^rs", names, invert = TRUE)   # other names   

underscore_pos = as.integer(regexpr("_", names[ind1], fixed = TRUE))   
names[ind1] = ifelse(underscore_pos > 0, substr(names[ind1], 1, underscore_pos - 1), names[ind1]) # rs1234567_A_C ==> rs1234567 
urls = paste0("http://www.phenoscanner.medschl.cam.ac.uk/?query=", names, "&catalogue=GWAS&p=1e-5&proxies=None&r2=0.8&build=37")  

pheno_links = character(nrow(results))
pheno_links[ind1] = paste0('<a href="', urls[ind1] ,'">', names[ind1] ,"</a>")
pheno_links[ind2] = names[ind2]     
results[,1] = pheno_links                   # use DT:datatable with escape = -2

chroms = results[,2]
positions = results[,3]
p1 = positions - 2000  
p2 = positions + 2000
urls = paste0("http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr", chroms, ":", p1, "-", p2)
marker_links = paste('<a href="', urls ,'">', positions ,"</a>", sep="")
results[,3] = marker_links 

cojo_neargene_file = paste(workfolder, cojo_neargene_file, sep ="/")
if(file.exists(cojo_neargene_file)) { 
  cojo_near_genes = get(load(cojo_neargene_file))   
} else {
  stop(paste("\n\n  ERROR (gwas_report.Rmd): File", cojo_neargene_file, "not found.\n\n"))
}
if(length(cojo_near_genes) != nrow(results)) stop(paste("\n\n  ERROR (gwas_report.Rmd): Inconsistent data", cojo_neargene_file, "vs.", cojo_file, "\n\n" )) 

genes = unlist(lapply(cojo_near_genes, `[[`, 1))                                             
chromosomes = unlist(lapply(cojo_near_genes, `[[`, 2))   
starts = unlist(lapply(cojo_near_genes, `[[`, 3)) 
ends = unlist(lapply(cojo_near_genes, `[[`, 4))
urls = paste0("http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr", chromosomes, ":", starts, "-", ends)    
gene_links = paste('<a href="', urls ,'">', genes ,"</a>", sep="")  
results = cbind(results, genes)
results[,ncol(results)] = gene_links  

results$P = signif(results$P, 4)  # no problem if NA
results$A1_FREQ = signif(results$A1_FREQ, 3)
results$BETA = signif(results$BETA, 4)
results$SE = signif(results$SE, 3)
```

<!-- need extra chunk for data table -->

```{r cojo_table, echo = FALSE, warning = FALSE, eval = cojo_done}
options(htmlwidgets.TOJSON_ARGS = list(na = 'string'))   # show NA (no empty fields!) 
datatable(results, class = 'cell-border stripe', escape = c(-2, -4, -11), width = 1200) %>% formatStyle('BETA', color = 'green') 
```





```{r cojo_orig_results, echo = FALSE, warning = FALSE, eval = cojo_done}
asis_output("#### Table with corresponding original results\\n")

cojo_orig_file = paste(workfolder, cojo_orig_file, sep ="/")
if(file.exists(cojo_orig_file)) {
  results = get(load(cojo_orig_file))
} else {
  stop(paste("\n\n  ERROR (gwas_report.Rmd): File", cojo_orig_file, "not found.\n\n"))
}
results = results[,-7]  # redundant 
rownames(results) = 1:nrow(results)  

# now, we can use the results from the chunk above ("cojo_results") because only P, beta, and SE are changed

results[,1] = pheno_links               # pheno_links, same as above     use DT:datatable with escape = -2
results[,3] = marker_links 		# marker links, same as above
results = cbind(results, genes)         # genes, same as above
results[,ncol(results)] = gene_links    # gene links, same as above
results$P = signif(results$P, 4)  
results$A1_FREQ = signif(results$A1_FREQ, 3)
results$BETA = signif(results$BETA, 4)
results$SE = signif(results$SE, 3)
```

<!-- need extra chunk for data table -->

```{r cojo_orig_table, echo = FALSE, warning = FALSE, eval = cojo_done}   
datatable(results, class = 'cell-border stripe', escape = c(-2, -4, -11), width = 1200) %>% formatStyle('BETA', color = 'green') 
```




```{r clump_results, echo = FALSE, warning = FALSE, eval = clump_done}
asis_output("#### Clump results\\n")

clump_file = paste(workfolder, clump_file, sep ="/")
if(file.exists(clump_file)) {
  results = get(load(clump_file))
} else {
  stop(paste("\n\n  ERROR (gwas_report.Rmd): File", clump_file, "not found.\n\n"))
}
results = results[,-7]  # redundant - remove 
rownames(results) = 1:nrow(results)  

names = results[,1]                        # all markers names 
ind1 = grep("^rs", names)                  # rs names               
ind2 = grep("^rs", names, invert = TRUE)   # other names   

underscore_pos = as.integer(regexpr("_", names[ind1], fixed = TRUE))   
names[ind1] = ifelse(underscore_pos > 0, substr(names[ind1], 1, underscore_pos - 1), names[ind1]) # rs1234567_A_C ==> rs1234567 
urls = paste0("http://www.phenoscanner.medschl.cam.ac.uk/?query=", names, "&catalogue=GWAS&p=1e-5&proxies=None&r2=0.8&build=37")  

pheno_links = character(nrow(results))
pheno_links[ind1] = paste0('<a href="', urls[ind1] ,'">', names[ind1] ,"</a>")
pheno_links[ind2] = names[ind2]     
results[,1] = pheno_links                   # use DT:datatable with escape = -2

chroms = results[,2]
positions = results[,3]
p1 = positions - 2000  
p2 = positions + 2000
urls = paste0("http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr", chroms, ":", p1, "-", p2)
marker_links = paste('<a href="', urls ,'">', positions ,"</a>", sep="")
results[,3] = marker_links 

clump_neargene_file = paste(workfolder, clump_neargene_file, sep ="/")
if(file.exists(clump_neargene_file)) { 
  clump_near_genes = get(load(clump_neargene_file))   
} else {
  stop(paste("\n\n  ERROR (gwas_report.Rmd): File", clump_neargene_file, "not found.\n\n"))
}
if(length(clump_near_genes) != nrow(results)) stop(paste("\n\n  ERROR (gwas_report.Rmd): Inconsistent data", clump_neargene_file, "vs.", clump_file, "\n\n" )) 

genes = unlist(lapply(clump_near_genes, `[[`, 1))                                             
chromosomes = unlist(lapply(clump_near_genes, `[[`, 2))   
starts = unlist(lapply(clump_near_genes, `[[`, 3)) 
ends = unlist(lapply(clump_near_genes, `[[`, 4))
urls = paste0("http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr", chromosomes, ":", starts, "-", ends)    
gene_links = paste('<a href="', urls ,'">', genes ,"</a>", sep="")  
results = cbind(results, genes)
results[,ncol(results)] = gene_links  

results$P = signif(results$P, 4)  
results$A1_FREQ = signif(results$A1_FREQ, 3)
results$BETA = signif(results$BETA, 4)
results$SE = signif(results$SE, 3)
```

<!-- need extra chunk for data table -->

```{r clump_table, echo = FALSE, warning = FALSE, eval = clump_done}   
datatable(results, class = 'cell-border stripe', escape = c(-2, -4, -11), width = 1200) %>% formatStyle('BETA', color = 'green') 
```


<!-- Show all significant markers: only if neither cojo nor clump was conducted --> 

```{r show_signif, echo = FALSE, warning = FALSE}
show_signif = !cojo_done & !clump_done
```



```{r signif_results, echo = FALSE, warning = FALSE, eval = show_signif}
asis_output("#### Unpruned significant markers\\n")

signif_file = paste(workfolder, signif_file, sep ="/")  
if(file.exists(signif_file)) {
  results = get(load(signif_file))
} else {
  stop(paste("\n\n  ERROR (gwas_report.Rmd): File", signif_file, "not found.\n\n"))
}
results = results[,-7]  # redundant
rownames(results) = 1:nrow(results)  

names = results[,1]                         
ind1 = grep("^rs", names)                                 
ind2 = grep("^rs", names, invert = TRUE)      

underscore_pos = as.integer(regexpr("_", names[ind1], fixed = TRUE))   
names[ind1] = ifelse(underscore_pos > 0, substr(names[ind1], 1, underscore_pos - 1), names[ind1]) # rs1234567_A_C ==> rs1234567 
urls = paste0("http://www.phenoscanner.medschl.cam.ac.uk/?query=", names, "&catalogue=GWAS&p=1e-5&proxies=None&r2=0.8&build=37")  

pheno_links = character(nrow(results))
pheno_links[ind1] = paste0('<a href="', urls[ind1] ,'">', names[ind1] ,"</a>")
pheno_links[ind2] = names[ind2]     
results[,1] = pheno_links                   

chroms = results[,2]
positions = results[,3]
p1 = positions - 2000  
p2 = positions + 2000
urls = paste0("http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr", chroms, ":", p1, "-", p2)
marker_links = paste('<a href="', urls ,'">', positions ,"</a>", sep="")
results[,3] = marker_links 

signif_neargene_file = paste(workfolder, signif_neargene_file, sep ="/")
if(file.exists(signif_neargene_file)) { 
  signif_near_genes = get(load(signif_neargene_file))   
} else {
  stop(paste("\n\n  ERROR (gwas_report.Rmd): File", signif_neargene_file, "not found.\n\n"))
}
if(length(signif_near_genes) != nrow(results)) stop(paste("\n\n  ERROR (gwas_report.Rmd): Inconsistent data", signif_neargene_file, "vs.", signif_file, "\n\n" )) 

genes = unlist(lapply(signif_near_genes, `[[`, 1))                                             
chromosomes = unlist(lapply(signif_near_genes, `[[`, 2))   
starts = unlist(lapply(signif_near_genes, `[[`, 3)) 
ends = unlist(lapply(signif_near_genes, `[[`, 4))
urls = paste0("http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr", chromosomes, ":", starts, "-", ends)    
gene_links = paste('<a href="', urls ,'">', genes ,"</a>", sep="")  
results = cbind(results, genes)
results[,ncol(results)] = gene_links  

results$P = signif(results$P, 4)  
results$A1_FREQ = signif(results$A1_FREQ, 3)
results$BETA = signif(results$BETA, 4)
results$SE = signif(results$SE, 3)
```
<!-- need extra chunk for data table -->


```{r signif_table, echo = FALSE, warning = FALSE, eval = show_signif}   
datatable(results, class = 'cell-border stripe', escape = c(-2, -4, -11), width = 1200) %>% formatStyle('BETA', color = 'green') 
```



## QQ-plot for $p$-values

```{r, echo=FALSE, fig.cap="", out.width = '100%', fig.align='left'}
plot = paste(workfolder, qq_plot, sep ="/")  
knitr::include_graphics(plot)
```


## Manhattan plot

```{r, echo=FALSE, fig.cap="", out.width = '100%', fig.align='left'}
plot = paste(workfolder, man_plot, sep ="/") 
knitr::include_graphics(plot)
```


## Histogram of $\beta$ values 

```{r, echo=FALSE, fig.cap="", out.width = '100%', fig.align='left'}
plot = paste(workfolder, histo_plot, sep ="/")
knitr::include_graphics(plot)
```


## Kernel density plot of $\beta$ values 

```{r, echo=FALSE, fig.cap="", out.width = '100%', fig.align='left'}
plot = paste(workfolder, kernel_plot, sep ="/")
knitr::include_graphics(plot)
```

<br> 



## Technical information  

 - GWAS workfolder: `r workfolder`

<br> 



<!-- uwemenzel@gmail.com -->







